File: dsa-interview-crash-course/1004_max_consecutive_ones_iii.py
from typing import List


class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        l = 0
        max_width = 0
        for i in range(len(nums)):
            if (nums[i] == 0):
                k -= 1
            
            if k < 0:
                if (nums[l] == 0):
                    k += 1
                l += 1
            
            max_width = max(max_width, i - l + 1)
            
        return max_width

File: dsa-interview-crash-course/1004_max_consecutive_ones_iii.cpp
#include <vector>

class Solution {
   public:
    int longestOnes(std::vector<int> &nums, int k) {
        int l = 0;
        int max_width = 0;

        for (int i = 0; i < nums.size(); ++i) {
            if (!nums[i]) {
                --k;
            }
            if (k < 0) {
                if (!nums[l]) {
                    ++k;
                }
                ++l;
            }
            max_width = std::max(max_width, i - l + 1);
        }
        return max_width;
    }
};